### PR TITLE
Add abc module from CPython

### PR DESCRIPTION
**Pull request by skinny121**
This copies over the pure python abc module implementation from CPython, with slight modifications. This is the first step in the Python side of the standard library.

Changes required to abc Python code:
- Changed `super().__new__` with `type.__new__` to workaround #778
- Changed a fronzenset to a set, to work around fronzenset not currently being iterable

Changes made to RustPython to make abc work:
- Made object hashable by default
- Added callback parameter to weakref.ref, which currently doesn't do anything, except not complain if it is passed.

I also copied over the PSF license file, so that we are complying with the terms.

### CODE REVIEW

**Comment by adrian17** [comment]
I kinda assumes we were to use https://github.com/beeware/ouroboros for starters (and just fill in core lower-level stdlib till it starts working)?

File: tests/snippets/test_abc.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+import abc
```
Comment (windelbouwman) [inline tests/snippets/test_abc.py]: Did you copy this file from cpython? Please note that cpython has its own test suite for the standard library. We should probably pick that one.

Comment (skinny121) [reply tests/snippets/test_abc.py]: Yes I did copy it from cpython. Eventually we should use cpython test suite, but that requires unittest to work, which It won't at this stage.

**Comment by skinny121** [comment]
@adrian17 I think it's useful to start to test core stdlib modules, to check where we aren't quite compatible yet, and ensure we maintain compatibility. We could do this by downloading/installing the stdlib during building, but this would mean we can't patch the stdlib which we probably need to.

**Comment by coolreader18** [comment]
I think ideally we shouldn't need to make patches, but that's probably not feasible at this point in development. Maybe we could have it as an optional test on Travis?

**Comment by adrian17** [comment]
Personally, I was hopeful that it would be possible to start trying to use ouroboros/CPython stdlib/CPython test suite, verbatim; obviously, 99% of code would fail on imports, but that would make a good point of reference to start filling in the blanks.
(and some of CPython test files could be run verbatim already, with just a basic `unittest` skeleton implementation)

File: tests/snippets/test_abc.py (hunk 1 above)
Comment (windelbouwman) [reply tests/snippets/test_abc.py]: unittest is an excellent candidate to port first. This I tried some time back, but too many things were broken at that time.

**Comment by windelbouwman** [comment]
My suggestion would be to make the unittest library work as first library. I'm fine with using ouroboros, using cpython or re-implementing all of the library. The main issue with ouroboros and cpython stdlib is the libraries which are half python, half c. Those are the weird ones. The pure python ones, we can simply copy paste.

File: PSF-LICENSE [hunk 2]
```diff
@@ -0,0 +1,254 @@
+A. HISTORY OF THE SOFTWARE
```
Comment (windelbouwman) [inline PSF-LICENSE]: Maybe this license file can be placed in the Lib folder, together with a readme indicating what it applies to.

**Comment by skinny121** [comment]
I agree unittest should be the first library that we target.

**Pull request merged**